### PR TITLE
Exclude string type annotations from ESP

### LINKED ISSUE #3435: Crash in preview style with multiline quoted return-position type hint
**Describe the bug**

When using `--preview`, this input causes error message "Unable to match a closing bracket to the following opening bracket: )". No issues when running without --preview

Looks similar to #3414, but I can still reproduce this issue in main branch.

**To Reproduce**

```python
from typing import Generic, TypeVar

T = TypeVar("T")


class VeryLongClassNameWithAwkwardGenericSubtype(Generic[T]):
    pass


def something(
    argument: (
        "VeryLongClassNameWithAwkwardGenericSubtype[int] |"
        "VeryLongClassNameWithAwkwardGenericSubtype[str]"
    ),
) -> (
    "VeryLongClassNameWithAwkwardGenericSubtype[int] |"
    "VeryLongClassNameWithAwkwardGenericSubtype[str]"
):
    ...
```

**Environment**

- Black's version: 22.12.1.dev7+gabd2b25
- OS and Python version: 3.10.8, macOS

### LINKED ISSUE #3457: Black --preview (ESP) crashes with implicitly concatenated stringified generic type params
**Describe the bug**

**To Reproduce**

For example, take this code:

```python
def func(
    arg1,
    arg2,
) -> Set["this_is_a_very_long_module_name.AndAVeryLongClasName"
         ".WithASubClass"]:
  pass
```

And run it with these arguments:

```sh
$ black --preview t.py
```

The resulting error is:

```
error: cannot format /Users/yileiyang/Downloads/t.py: Unable to match a closing bracket to the following opening bracket: )

Oh no! 💥 💔 💥
1 file would fail to reformat.
```

**Expected behavior**

**Environment**

- Black's version: this crash existed since v22.1.0, doesn't affect v23.1a1 since it's ESP

**Additional context**

```
Traceback (most recent call last):
  File "/Users/yileiyang/github.com/yilei/black/src/black/brackets.py", line 99, in mark
    opening_bracket = self.bracket_match.pop((self.depth, leaf.type))
KeyError: (-1, 8)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/yileiyang/github.com/yilei/black/src/black/__init__.py", line 1077, in format_str
    dst_contents = _format_str_once(src_contents, mode=mode)
  File "/Users/yileiyang/github.com/yilei/black/src/black/__init__.py", line 1109, in _format_str_once
    for line in transform_line(
  File "/Users/yileiyang/github.com/yilei/black/src/black/linegen.py", line 577, in transform_line
    result = run_transformer(line, transform, mode, features, line_str=line_str)
  File "/Users/yileiyang/github.com/yilei/black/src/black/linegen.py", line 1420, in run_transformer
    result.extend(transform_line(transformed_line, mode=mode, features=features))
  File "/Users/yileiyang/github.com/yilei/black/src/black/linegen.py", line 577, in transform_line
    result = run_transformer(line, transform, mode, features, line_str=line_str)
  File "/Users/yileiyang/github.com/yilei/black/src/black/linegen.py", line 1416, in run_transformer
    for transformed_line in transform(line, features):
  File "/Users/yileiyang/github.com/yilei/black/src/black/trans.py", line 251, in __call__
    for line_result in self.do_transform(line, string_idx):
  File "/Users/yileiyang/github.com/yilei/black/src/black/trans.py", line 395, in do_transform
    msg_result = self._merge_string_group(new_line, string_idx)
  File "/Users/yileiyang/github.com/yilei/black/src/black/trans.py", line 602, in _merge_string_group
    append_leaves(new_line, line, [leaf])
  File "/Users/yileiyang/github.com/yilei/black/src/black/lines.py", line 707, in append_leaves
    new_line.append(new_leaf, preformatted=preformatted)
  File "/Users/yileiyang/github.com/yilei/black/src/black/lines.py", line 81, in append
    self.bracket_tracker.mark(leaf)
  File "/Users/yileiyang/github.com/yilei/black/src/black/brackets.py", line 101, in mark
    raise BracketMatchError(
black.brackets.BracketMatchError: Unable to match a closing bracket to the following opening bracket: )
```

### PR DESCRIPTION
**Pull request by yilei**
### Description

This PR excludes string type annotations from ESP, and also fixes crash on mismatched brackets from ESP.

1. Unfortunately pyright doesn't support stringified annotations that span multiple lines, e.g.:

```python
def something(
    arg,
) -> (
    "VeryLongClassNameWithAwkwardGenericSubtype[int] |"
    "VeryLongClassNameWithAwkwardGenericSubtype[str]"
):
    ...
```

Even though other type checkers (mypy, pytype, pyre) do support them. To be safe, we can choose to exclude them from ESP for now. Note we _could_ technically choose to try merging them into a single string when the result fits in the line-length, but this would be more complicated than simply excluding them from ESP.

2. The crash is because of unmatched brackets. In ESP, it can process lines continued from earlier line breaks, e.g.:

```python
def something(
    arg,
) -> "VeryLongClassNameWithAwkwardGenericSubtype[int] | VeryLongClassNameWithAwkwardGenericSubtype[str]":
    ...
```

The line can be `) -> "VeryLongClassNameWithAwkwardGenericSubtype[int] |VeryLongClassNameWithAwkwardGenericSubtype[str]":` while processing. This PR makes `BracketTracker.mark` accept unmatched closing brackets. 

Fixes #3435. Fixes #3457.

### Checklist - did you ...

- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

File: src/black/brackets.py [hunk 1]
```diff
@@ -80,9 +76,12 @@ def mark(self, leaf: Leaf) -> None:
         within brackets a given leaf is. 0 means there are no enclosing brackets
         that started on this line.
 
-        If a leaf is itself a closing bracket, it receives an `opening_bracket`
-        field that it forms a pair with. This is a one-directional link to
-        avoid reference cycles.
+        If a leaf is itself a closing bracket and there is a matching opening
+        bracket earlier, it receives an `opening_bracket` field that it forms a
```
Comment (JelleZijlstra) [inline src/black/brackets.py:80]:
```suggestion
        bracket earlier, it receives an `opening_bracket` field with which it forms a
```

File: src/black/brackets.py [hunk 2]
```diff
@@ -91,17 +90,18 @@ def mark(self, leaf: Leaf) -> None:
...
         if leaf.type in CLOSING_BRACKETS:
             self.depth -= 1
-            try:
-                opening_bracket = self.bracket_match.pop((self.depth, leaf.type))
-            except KeyError as e:
-                raise BracketMatchError(
-                    "Unable to match a closing bracket to the following opening"
-                    f" bracket: {leaf}"
-                ) from e
+            opening_bracket = self.bracket_match.pop((self.depth, leaf.type))
```
Comment (JelleZijlstra) [inline src/black/brackets.py:104]: Why not keep the BracketMatchError? Seems like it makes for a better error message.

**Review by JelleZijlstra — COMMENTED** [review]
Thanks!

File: src/black/brackets.py (hunk 2 above)
Comment (yilei) [reply src/black/brackets.py]: Good point, done.

**Pull request merged**